plugins/s3_auth/aws_auth_v4.cc

```diff
@@ -0,0 +1,730 @@
+       * 'A'-'Z', 'a'-'z', '0'-'9', '-', '.', '_', and '~'. */
+      result << in[i];
+    } else if (in[i] == ' ') {
+      /* The space character is a reserved character and must be encoded as "%20" (and not as "+"). */
+      result << "%20";
+    } else if (isObjectName && in[i] == '/') {
+      /* Encode the forward slash character, '/', everywhere except in the object key name. */
+      result << "/";
+    } else {
+      /* Letters in the hexadecimal value must be upper-case, for example "%1A". */
+      result << "%" << std::uppercase << std::setfill('0') << std::setw(2) << std::hex << (int)in[i];
+    }
+  }
+
+  return result.str();
+}
+
+/**
+ * @brief lower case a C string
+ *
+ * @param in ptr to the char string to be converted
+ * @param inLen length of the string to be converted
+ * @return a string converted to lower case.
+ */
+String
+lowercase(const char *in, size_t inLen)
+{
+  if (nullptr == in || inLen == 0) {
+    return String();
+  }
+
+  std::stringstream result;
+
+  std::locale loc;
+  for (std::string::size_type i = 0; i < inLen; ++i) {
+    result << std::tolower(*(in + i), loc);
+  }
+
+  return result.str();
+}
+
+/**
+ * @brief lower case a C++ string
+ *
+ * @param in ptr to the char string to be converted
+ * @param inLen length of the string to be converted
+ * @return a string converted to lower case.
+ */
+String
+lowercase(const String &in)
+{
+  std::stringstream result;
+
+  std::locale loc;
+  for (auto i : in) {
+    result << std::tolower((i), loc);
+  }
+
+  return result.str();
+}
+
+/**
+ * @brief trim the white-space character from the beginning and the end of the string ("in-place", just moving pointers around)
+ *
+ * @param in ptr to an input string
+ * @param inLen input character count
+ * @param newLen trimmed string character count.
+ * @return pointer to the trimmed string.
+ */
+const char *
+trimWhiteSpaces(const char *in, size_t inLen, size_t &newLen)
+{
+  if (nullptr == in || inLen == 0) {
+    return in;
+  }
+
+  const char *first = in;
+  while (size_t(first - in) < inLen && isspace(*first)) {
+    first++;
+  }
+
+  const char *last = in + inLen - 1;
+  while (last > in && isspace(*last)) {
+    last--;
+  }
+
+  newLen = last - first + 1;
+  return first;
+}
+
+/**
+ * @brief Trim white spaces from beginning and end.
+ * @returns trimmed string
+ */
+String
+trimWhiteSpaces(const String &s)
+{
+  /* @todo do this better? */
+  static const String whiteSpace = " \t\n\v\f\r";
+  size_t start                   = s.find_first_not_of(whiteSpace);
+  if (String::npos == start) {
+    return String();
+  }
+  size_t stop = s.find_last_not_of(whiteSpace);
+  return s.substr(start, stop - start + 1);
+}
+
+/*
+ * Group of static inline helper function for less error prone parameter handling and unit test logging.
+ */
+inline static void
+sha256Update(SHA256_CTX *ctx, const char *in, size_t inLen)
+{
+  SHA256_Update(ctx, in, inLen);
+#ifdef AWS_AUTH_V4_DETAILED_DEBUG_OUTPUT
```

Comment:
Same here, naming convention. I'm ok either way, but I think if we're going to call this "aws" now instead of "s3", we should be consistent. So, S3_AUTH_V4_DETAILED_DEBUG_OUTPUT ?